README.md
---
deeptendies
---

[![image](https://img.shields.io/pypi/v/deeptendies.svg)](https://pypi.python.org/pypi/deeptendies)
[![image](https://img.shields.io/travis/stancsz/deeptendies.svg)](https://travis-ci.com/stancsz/deeptendies)
[![Documentation Status](https://readthedocs.org/projects/deeptendies/badge/?version=latest)](https://deeptendies.readthedocs.io/en/latest/?version=latest)

Bringing quantitative trading to the masses!

The most difficult part of quantitative analysis is getting started. This project has you covered ;)

It is a one-stop shop for obtaining historical data, engineering features, and fitting the data through a pipeline.


# Features
Let's open a blank colab and try out this library: https://colab.research.google.com/#create=true

```python
# 1 Liner easy install
!pip install git+https://github.com/stancsz/deeptendies && pip install -r https://raw.githubusercontent.com/stancsz/deeptendies/main/requirements.txt
import deeptendies as dt

# Look and feel of pandas usage & get a pd.DataFrame
df = dt.DataFrame.from_yf('GME')
print(type(df))

# Builtin Pipeline class for mass features processing
pipeline = dt.Pipeline(
    [
        dt.Feature.get_x_high,
        dt.Feature.get_x_low,
        dt.Feature.get_x_ma,
        dt.Feature.get_diff
    ]
)
df = pipeline.run(df=df, x=50, interval='day')
df = pipeline.run(df=df, x=100, interval='day')
df = pipeline.run(df=df, x=200, interval='day')
df = pipeline.run(df=df, x=13, interval='week')
df = pipeline.run(df=df, x=26, interval='week')
df = pipeline.run(df=df, x=52, interval='week')
df[['50_day_ma','200_day_ma']].plot()
```
happy quanting :)

## Additional Features: Feature engineering
Easy to use feature engineering methods
```python
df = dt.Feature.get_x_low(df, x=52, interval='week')
df = get_x_ma(df, x=50, interval='day')
```
learn more @ `deeptendies/feature.py`

![img.png](docs/img.png)
## Jupyter
[or use it in a notebook](https://github.com/deeptendies/deeptendies/blob/master/tests/jupyter/01_getting_stock_data_and_engineer_feature.ipynb)

# Development guide
```
git clone https://github.com/deeptendies/deeptendies.git
pip install -e deeptendies
```

# Credits
- This package is redesigned from the [legacy deeptendies package](https://github.com/deeptendies/legacy-deeptendies-library), credits to original authors.
  - @mklasby @bgulseren @KBehairy @hasnil @Karenzhang7717
- This package was created with Cookiecutter and the audreyr/cookiecutter-pypackage project template.

# License
- Free software: MIT license
- Documentation: <https://deeptendies.readthedocs.io>.


HISTORY.rst
=======
History
=======

0.1.0 (2022-03-30)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for deeptendies."""


tests/test_database.py
#!/usr/bin/env python
"""Tests for `deeptendies` package."""
import pytest
from unittest import TestCase

class DatabaseTestSuite(TestCase):
    def test_database(self):
        """
        test database module
        :return:
        """
        import deeptendies as dt
        db = dt.Database()
        result = db.execute("""CREATE TABLE test (
        test_id int,
        test_column_a varchar(50),
        test_column_b varchar(50));""")
        print(result)
        result = db.execute("""SELECT * FROM test;""")
        print(result)
        assert result is not None


    def test_db_save(self):
        import deeptendies as dt
        db = dt.Database()
        # method 1
        df = dt.DataFrame.from_yf('SQQQ', db=db)
        print(db.read("SQQQ").head())

        # method 2
        df = dt.DataFrame.from_yf('GME')
        db.save(df, 'GME')
        result = db.read("GME").head()
        print(result )
        assert result is not None


    def test_db_union(self):
        import deeptendies as dt
        db = dt.Database()
        # method 1
        dt.DataFrame.from_yf('GME', db=db)
        dt.DataFrame.from_yf('TSLA', db=db)
        dt.DataFrame.from_yf('TQQQ', db=db)
        df = db.getall(['GME', 'TSLA', 'TQQQ'])
        print(df.head())
        print(df.tail())
        assert df is not None


    def test_backfill(self):
        import deeptendies as dt
        tickers = ['YINN', 'YANG', 'GME', 'TQQQ', 'SQQQ']
        df = dt.DataFrame.db_load(tickers)
        print(df)
        assert df is not None


tests/test_pipeline.py
#!/usr/bin/env python

"""Tests for `deeptendies` package."""
import pytest

import pytest
from unittest import TestCase


class PipelineTestSuite(TestCase):
    @pytest.fixture
    def from_yf_info(self):
        """
        pytest fixture to get yahoo finance info
        :return:
        """
        from deeptendies.data import DataFrame
        dt = DataFrame()
        df = dt.from_yf_info('TQQQ')
        print(type(df))
        print(df)

    @pytest.fixture
    def from_yf(self):
        """
        pytest fixture to get dataframe from yahoo finance
        :return:
        """
        import deeptendies as dt
        df = dt.DataFrame.from_yf('TQQQ')
        return df


    def test_pipeline(self, df=from_yf):
        """
        test pipeline class, as well as a few feature engineering operators. then assert the dataframe's column with
        the expected values.
        :return:
        """
        import deeptendies as dt
        pipeline = dt.Pipeline(
            [
                dt.Feature.get_x_high,
                dt.Feature.get_x_low,
                dt.Feature.get_x_ma,
                dt.Feature.get_diff
            ]
        )
        df = pipeline.run(df=df, x=50)
        df = pipeline.run(df=df, x=100)
        df = pipeline.run(df=df, x=200)
        df = pipeline.run(df=df, x=50, interval='week')
        df = pipeline.run(df=df, x=100, interval='week')
        df = pipeline.run(df=df, x=200, interval='week')
        print(df.head())
        print(df.columns)
        expected = ['high', 'low', 'open', 'close', 'volume', 'adj_close', '50_day_high',
                    '50_day_low', '50_day_ma', 'high_diff', 'low_diff', 'close_diff',
                    'volume_diff', '100_day_high', '100_day_low', '100_day_ma',
                    '200_day_high', '200_day_low', '200_day_ma', '50_week_high',
                    '50_week_low', '50_week_ma', '100_week_high', '100_week_low',
                    '100_week_ma', '200_week_high', '200_week_low', '200_week_ma']
        assert all([a == b for a, b in zip(df.columns, expected)])  # https://codefrom.page.link/RtQw


deeptendies/data.py
import datetime

from yfinance import tickers

import deeptendies
import yfinance as yf
import pandas as pd
import json
from datetime import date
from pandas_datareader import data as pdr

pd.options.plotting.backend = 'plotly'
"""Main module."""


class DataFrame:
    def __init__(self, *args, **kwargs):
        yf.pdr_override()

    @staticmethod
    def from_yf(ticker, start=date.today() - datetime.timedelta(days=10 * 365), end=date.today(), db=None):
        """returns a pandas dataframe containing the last 10 years of data"""
        df = pdr.get_data_yahoo(ticker, start=start, end=end)
        df.columns = [x.lower().replace(' ', '_') for x in df.columns]
        if db is not None:
            db.save(df=df, name=ticker.lower())
        return df

    @staticmethod
    def from_yf_info(ticker, db=None):
        yft = yf.Ticker(ticker)
        df = pd.DataFrame([yft.info], columns=yft.info.keys()).T
        df.columns = ['value']
        if db is not None:
            db.save(df=df, name=f'{ticker}_info'.lower())
        return df

    @staticmethod
    def db_load(tickers):
        db = deeptendies.Database()
        for t in tickers:
            deeptendies.DataFrame.from_yf(t, db=db)
        return db.getall(tickers)


deeptendies/backend.py
import sqlite3
import pandas as pd


class Database():
    def __init__(self, *args, **kwargs):
        try:
            self.conn = sqlite3.connect('sqlite.db')
        except Exception as e:
            print(e)
        print('sqlite.db connection established')

    def execute(self, query, *args, **kwargs):
        query_cursor = self.conn.cursor()
        query_cursor.execute(query, *args, **kwargs)
        return query_cursor.fetchall()

    def save(self, df, name):
        df.to_sql(name.lower(), self.conn, if_exists='replace')

    def read(self, name):
        return pd.read_sql(f'SELECT * FROM {name.lower()};', con=self.conn)

    def getall(self, tickers):
        union_block = """SELECT '{}' ticker, * FROM {}"""
        blocks = [union_block.format(x.lower(), x.lower()) for x in tickers]
        query = ' UNION ALL '.join(blocks)
        return pd.read_sql(query, con=self.conn)


deeptendies/feature.py
class Feature:
    def __init__(self, *args, **kwargs):
        pass

    @staticmethod
    def get_x_low(df, x=52, interval='day', **kwargs):
        if interval == 'week':
            x_inter = x * 7
        elif interval == 'day':
            x_inter = x
        df[f'{x}_{interval}_low'] = df['low'].rolling(window=x_inter, min_periods=x_inter).min()
        return df

    @staticmethod
    def get_x_high(df, x=52, interval='day', **kwargs):
        if interval == 'week':
            x_inter = x * 7
        elif interval == 'day':
            x_inter = x
        df[f'{x}_{interval}_high'] = df['high'].rolling(window=x_inter, min_periods=x_inter).max()
        return df

    @staticmethod
    def get_x_ma(df, x=50, interval='day', **kwargs):
        if interval == 'week':
            x_inter = x * 7
        elif interval == 'day':
            x_inter = x
        df[f'{x}_{interval}_ma'] = df['close'].rolling(window=x_inter, min_periods=x_inter).mean()
        return df

    @staticmethod
    def get_diff(df, cols=['high', 'low', 'close', 'volume'], **kwargs):
        for col in cols:
            df[f'{col}_diff'] = df[col].diff()
        return df
